design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+      timeout: 30s
+    # An x-kubernetes-embedded-resource RawExtension (i.e. an unschemafied)
+    # Kubernetes resource. Passed to the function as the functionConfig block of
+    # its ResourceList.
+    config:
+      apiVersion: database.example.org/v1alpha1
+      kind: Config
+      metadata:
+        name: cloudsql
+      spec:
+        version: POSTGRES_9_6
+```
+
+### Function API
+
+This document proposes that each function uses a `ResourceList` type derived
+from the [KRM function specification][krm-fn-spec] as its input and output.
+
+```yaml
+apiVersion: fn.apiextensions.crossplane.io/v1
+kind: ResourceList
+functionConfig:
+  apiVersion: database.example.org/v1alpha1
+  kind: Config
+  metadata:
+    name: cloudsql
+  spec:
+    version: POSTGRES_9_6
+items:
+- apiVersion: database.example.org/v1alpha1
+  kind: XPostgreSQLInstance
+  metadata:
+    name: my-db
+    annotations:
+      fn.apiextensions.crossplane.io/type: "CompositeResource"
+  spec:
+    parameters:
+      storageGB: 20
+    compositionSelector:
+      matchLabels:
+        provider: gcp
+    writeConnectionSecretToRef:
+      name: db-conn
+```
+
+Like a KRM function, a Composition function uses a `ResourceList` that consists
+of the following top-level fields:
+
+* The `apiVersion` and `kind` (required).
+* A `functionConfig` (optional). This is a [Kubernetes resource][rawextension]
+  with an arbitrary schema that may be used to provide additional configuration
+  to a function. For example a `render-helm-chart` function might use its
+  `functionConfig` to specify which Helm chart to render.
+* An `items` array (required). The XR and zero or more composed resources.
+* A `results` object (optional). Used to communicate information about the
+  result of a function, including warnings and errors.
+
+Each function in the array would be executed as a pipeline, with each function:
+
+1. Taking the XR and zero or more composed resources as input.
+1. Producing the optionally mutated XR, optionally mutated composed resources,
+   and newly rendered composed resources as output.
+
+This allows the output of one function to be the input to the next. The first
+function in the pipeline would be supplied with either:
+
+1. Only the XR. This would be the case for newly created XRs whose Composition
+   did not include a `resources` array.
+1. The XR and one or more resources. This would be the case if either:
+   1. The Composition included both a `resources` array and a `functions` array.
+      In this case the `resources` array is computed first then passed to
+      `functions`.
+   1. This was not a newly created XR, but instead an XR that had already
+      created its composed resources and was now being updated.
+
+In the case of `Container` functions this input and output would correspond to
+stdin and stdout. This may differ for future function implementations; for
+example a hypothetical webhook function implementation might take the input as
```

Comment:
This is also a bit confusing, why wouldn't the main kind of function be HTTP based? It looks like by default the Container function reads from stdin and write to the stdout. I believe that having a webhook approach by default will make everything easier as there are tools solving 98% of all the needs here.

Reply:
See comment above.  There's a lot less involved to package function logic as an OCI containerized process than there is in spinning up something to serve webhooks.